File: Algorithms/C++/00021-Merge Two Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if(list1 == nullptr && list2 == nullptr) return nullptr;
        if(list2 == nullptr && list1 != nullptr) return list1;
        if(list1 == nullptr && list2 != nullptr) return list2;

        int firstVal{0};
        if(list1->val <= list2->val){
            firstVal = list1->val;
            list1 = list1->next;
        } else{
            firstVal = list2->val;
            list2 = list2->next;
        }

        ListNode* sorted = new ListNode(firstVal);
        ListNode* result = sorted;

        while(!(list1 == nullptr && list2 == nullptr)){
            // if both lists have nodes left
            if(list1 != nullptr && list2 != nullptr){
                if(list1->val <= list2->val){
                    ListNode* node = new ListNode(list1->val);
                    sorted->next = node;
                    sorted = sorted->next;
                    list1 = list1->next;
                } else{
                    ListNode* node = new ListNode(list2->val);
                    sorted->next = node;
                    sorted = sorted->next;
                    list2 = list2->next;
                }
            }

            // if only one list has nodes left
            if(list1 == nullptr && list2 != nullptr){
                while(list2 != nullptr){
                    ListNode* node = new ListNode(list2->val);
                    sorted->next = node;
                    sorted = sorted->next;
                    list2 = list2->next;
                }
            } else if(list2 == nullptr && list1 != nullptr){
                
                while(list1 != nullptr){
                    ListNode* node = new ListNode(list1->val);
                    sorted->next = node;
                    sorted = sorted->next;
                    list1 = list1->next;
                }
            }

        }

        return result;

    }
};

File: Algorithms/Python3/00021-Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        newList = dummy = ListNode()

        # merge nodes up to length of shorter list
        while list1 and list2:
            if list1.val < list2.val:
                value = list1.val
                list1 = list1.next
            else:
                value = list2.val
                list2 = list2.next
            node = ListNode(value)
            newList.next = node
            newList = newList.next

        # merge rest of nodes from longer list
        while list1 or list2:
            if list1:
                value = list1.val
                list1 = list1.next
            elif list2:
                value = list2.val
                list2 = list2.next
            node = ListNode(value)
            newList.next = node
            newList = newList.next 
                  
        return dummy.next

'''
    newList 
    [dummy, 1, 1, 2]

    [1,2,3,4,5]
    [1,2,3]

'''